### PR TITLE
Add CancellableMutex

### PR DESCRIPTION
**Pull request by mads256h**
This mutex owns the data it locks.
It can be cancelled when it waits for the lock.

### CODE REVIEW

File: CentralHub.Api.Tests/CancellableMutexTests.cs [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+            {
+                // This task tries to lock but we have already locked in the first line of this function.
+                // So after 100 milliseconds in the cancelTask we cancel and we should expect to get an OperationCanceledException.
+                Assert.Throws<OperationCanceledException>(() => _cancellableMutex.Lock(_ => { }, tokenSource.Token).GetAwaiter().GetResult());
+            }, default);
+
+            var cancelTask = Task.Run(async () =>
+            {
+                // This makes the test always at least take 100 milliseconds to complete.
+                // Too bad :)
```
Comment (Mast3rwaf1z) [inline CentralHub.Api.Tests/CancellableMutexTests.cs:37]: grr

**Review by Taoshix — APPROVED** [review]
Gotta love Mads and his Semaphores

File: CentralHub.Api.Tests/CancellableMutexTests.cs (hunk 1 above)
Comment (mads256h) [reply CentralHub.Api.Tests/CancellableMutexTests.cs:37]: Jeg kan ikke teste det på anderledes vis desværre

Comment (mads256h) [reply CentralHub.Api.Tests/CancellableMutexTests.cs:37]: 100 millisekunder er heller ikke meget

File: CentralHub.Api/Threading/CancellableMutex.cs [hunk 2]
```diff
@@ -0,0 +1,74 @@
+namespace CentralHub.Api.Threading;
+
+public sealed class CancellableMutex<T> : IDisposable
+{
+    private readonly SemaphoreSlim _semaphoreSlim = new SemaphoreSlim(1, 1);
+
+    private readonly T _innerObject;
+
+    public CancellableMutex(T innerObject)
```
Comment (Mast3rwaf1z) [inline CentralHub.Api/Threading/CancellableMutex.cs:9]: is the templating really needed? i've only seen it in this language

Comment (Mast3rwaf1z) [reply CentralHub.Api/Threading/CancellableMutex.cs:9]: ah i see, you need it for using lambda in the call to the lock right?

File: CentralHub.Api/Services/SampleRoomRepository.cs [hunk 3]
```diff
@@ -1,14 +1,18 @@
...
-    private static readonly Dictionary<int, RoomDto> Rooms = new Dictionary<int, RoomDto>();
-    private static int _nextId;
+    private sealed class LockedStuff
+    {
+        public Dictionary<int, RoomDto> Rooms { get; } = new Dictionary<int, RoomDto>();
+        public int NextId { get; set; }
+    }
 
-    private static readonly object LockObject = new object();
+    private static readonly CancellableMutex<LockedStuff> LockedStuffMutex = new CancellableMutex<LockedStuff>(new LockedStuff());
```
Comment (Mast3rwaf1z) [inline CentralHub.Api/Services/SampleRoomRepository.cs:15]: seems a little strange that you're locking two resources at the same time? are we never accessing one without accessing the other?

**Review by Mast3rwaf1z — APPROVED** [review]
looks fine, change things i mentioned if you want but i don't think its critical to the functionality

File: CentralHub.Api/Services/SampleRoomRepository.cs (hunk 3 above)
Comment (mads256h) [reply CentralHub.Api/Services/SampleRoomRepository.cs:15]: We did that before. I think it makes sense to lock both at the same time as when we need to add a new room we need to increment NextId and we need to add it to the dictionary which requires both to be locked anyway.

File: CentralHub.Api/Threading/CancellableMutex.cs (hunk 2 above)
Comment (mads256h) [reply CentralHub.Api/Threading/CancellableMutex.cs:9]: Yes you get the value in the lambda and when you are in the lambda the lock is held

**Pull request merged**